### ISSUE TITLE
dpath.util.new() should raise TypeError when user asks to create a new path in an existing but unsubscriptable object

### ISSUE THREAD

**Issue by jackmaney** [akesterson/dpath-python#47]
OS X 10.9.5
Tested on Python 3.5.0, 3.4.3, and 2.7.10

So, I'm trying to use dpath in one of my projects, and the examples in the README docs work just fine. However, when I fiddle around on my own, `dpath.util.new` seems to think I want to tack on an array when I really don't:

```
In [1]: import dpath.util

In [2]: d = {"a": 1, "b": 2}

In [3]: dpath.util.new(d, "a/c", "something")
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/Users/[username]/.pyenv/versions/3.5.0/lib/python3.5/site-packages/dpath/path.py in path_types(obj, path)
     29         try:
---> 30             result.append([path[-1], cur[path[-1]].__class__])
     31         except TypeError:

TypeError: 'int' object is not subscriptable

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-3-e13767a0cab0> in <module>()
----> 1 dpath.util.new(d, "a/c", "something")

/Users/[username]/.pyenv/versions/3.5.0/lib/python3.5/site-packages/dpath/util.py in new(obj, path, value, separator)
     44     """
     45     pathlist = __safe_path__(path, separator)
---> 46     pathobj = dpath.path.path_types(obj, pathlist)
     47     return dpath.path.set(obj, pathobj, value, create_missing=True)
     48

/Users/[username]/.pyenv/versions/3.5.0/lib/python3.5/site-packages/dpath/path.py in path_types(obj, path)
     30             result.append([path[-1], cur[path[-1]].__class__])
     31         except TypeError:
---> 32             result.append([path[-1], cur[int(path[-1])].__class__])
     33     except (KeyError, IndexError):
     34         result.append([path[-1], path[-1].__class__])

ValueError: invalid literal for int() with base 10: 'c'
```

I don't understand what the substantive difference is between this and [the analogous example in the docs](https://github.com/akesterson/dpath-python#example-adding-new-keys). Am I missing anything obvious, here?

**Comment by akesterson** [comment]
What version of dpath are you using?

**Comment by jackmaney** [comment]
1.4.0

Also, would it be possible to request the addition of a `__version__` attribute to the `dpath` module? That would make finding the version a fair amount easier. :)

**Comment by akesterson** [comment]
**TL;DR - You can't ask dpath.util.new() to create things inside of an integer, it's not a subscriptable type.**

> would it be possible to request the addition of a __version__

We already have one, in a slightly different place.

    >>> import dpath.version
    >>> dpath.version.VERSION
    '1.3-1'

As far as this bug report, this is primarily user error. You're trying to create new path entries inside of an integer, which is expected to fail (though I admit the way it's failing is confusing, and is really a bug in itself).

    d = {"a": 1, "b": 2}
    dpath.util.new(d, "a/c", "something")

You're telling dpath "inside the object (d), inside of the object at path "a/", create a new attribute "c" with the value of the string "something". In other words, "please make d['a']['c'] = 'something' ". But the object at path /a/ is not an array or a hash, so we can't create d['a']['c']. So we would expect to see some kind of TypeError or AttributeError failure here. But the one you're getting is unclear and indicates that dpath isn't detecting that /a/ is not an array or index, which is a bug in itself.

It's different from the example you referenced because, in the example you referenced, we're asking dpath.util.new() to create new objects inside of existing array or hash objects. Specifically we're asking it to create the path /a/b/e/f/g in the object which already has the existing path /a/b/, which is a nested dictionary. So dpath creates /a/b/e/ as a dictionary, so on and so on, until it creates the terminus node. This same thing happens with arrays, if we were to ask it to create a new path inside of an existing array for us.

dpath really should've emitted a more clear error about what you were trying to do, regardless. I'll leave this open for that purpose.

**Comment by calebcase** [comment]
With PR #50:

```
>>> d = {"a": 1, "b": 2}
>>> dpath.util.new(d, "a/c", "something")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "dpath/util.py", line 56, in new
    return dpath.segments.set(obj, segments, value)
  File "dpath/segments.py", line 326, in set
    raise PathNotFound('Path: {}[{}]'.format(segments, i))
dpath.exceptions.PathNotFound: Path: ['a', 'c'][0]
```
What do we think? Is this clear enough?

**Comment by akesterson** [comment]
PR #50 has been deferred to 2.0, but this (in my opinion) does satisfy the user need for a more explicit error message. Given the low reporting volume on this (and the fact that it's an error message issue), it can probably wait til 2.0. Closing this with the intention that it will be solved with 2.0.

**Comment by jackmaney** [comment]
Looks great. Thanks!
